File: mcp-server/src/secure-channel.ts
import { createCipheriv, createDecipheriv, hkdfSync, randomBytes, timingSafeEqual } from "node:crypto";

/**
 * Client half of the DroidPilot secure channel.
 *
 * This mirrors `SecureChannel.kt` on the device exactly, and the two are pinned together
 * by shared test vectors in `test/secure-channel.test.ts` and `SecureChannelTest.kt`: both
 * sides assert that a fixed (secret, salt, plaintext) triple produces one specific
 * ciphertext. A change to either implementation that breaks compatibility fails a test
 * rather than silently failing to connect in the field.
 *
 * See the Kotlin file for the rationale and the construction; the summary is HKDF-SHA256
 * for the key schedule, AES-256-GCM per record, a 4-byte direction prefix plus an 8-byte
 * counter for the nonce, and strictly increasing counters to reject replays.
 */

export const SECRET_BYTES = 32;
export const SALT_BYTES = 16;

const NONCE_BYTES = 12;
const IV_PREFIX_BYTES = 4;
// The remaining 8 nonce bytes are the record counter, written by `writeBigUInt64BE`.
const TAG_BYTES = 16;
const KEY_BYTES = 32;

/** Domain-separation label. Must match `SecureChannel.HKDF_INFO` on the device. */
const HKDF_INFO = "droidpilot/v2/keys";

interface DirectionKeys {
  key: Buffer;
  ivPrefix: Buffer;
}

export class SecureChannel {
  private sendCounter = 0n;
  private highestSeenReceiveCounter = -1n;

  private constructor(
    private readonly send: DirectionKeys,
    private readonly receive: DirectionKeys,
  ) {}

  /**
   * Derives a view of the channel from the pairing secret and the per-session salt.
   *
   * `role` defaults to `"client"`, which is what the MCP server always uses. The `"server"`
   * role exists so tests can stand up a fake device that speaks the real protocol — that
   * is how the handshake, authentication rejection and mid-session disconnect paths are
   * covered without a physical phone.
   */
  static derive(secret: Buffer, salt: Buffer, role: "client" | "server" = "client"): SecureChannel {
    if (secret.length !== SECRET_BYTES) {
      throw new Error(`Pairing secret must be ${SECRET_BYTES} bytes, got ${secret.length}`);
    }
    if (salt.length !== SALT_BYTES) {
      throw new Error(`Session salt must be ${SALT_BYTES} bytes, got ${salt.length}`);
    }

    const total = 2 * (KEY_BYTES + IV_PREFIX_BYTES);
    const material = Buffer.from(
      hkdfSync("sha256", secret, salt, Buffer.from(HKDF_INFO, "utf8"), total),
    );

    let offset = 0;
    const take = (n: number): Buffer => {
      const slice = material.subarray(offset, offset + n);
      offset += n;
      return slice;
    };

    const c2sKey = take(KEY_BYTES);
    const c2sIv = take(IV_PREFIX_BYTES);
    const s2cKey = take(KEY_BYTES);
    const s2cIv = take(IV_PREFIX_BYTES);

    // The client sends on c2s and receives on s2c; the device's view is the mirror image.
    return role === "client"
      ? new SecureChannel({ key: c2sKey, ivPrefix: c2sIv }, { key: s2cKey, ivPrefix: s2cIv })
      : new SecureChannel({ key: s2cKey, ivPrefix: s2cIv }, { key: c2sKey, ivPrefix: c2sIv });
  }

  /** Encrypts `plaintext` into a self-contained record: nonce ‖ ciphertext ‖ tag. */
  seal(plaintext: Buffer): Buffer {
    const counter = this.sendCounter;
    this.sendCounter += 1n;

    const nonce = buildNonce(this.send.ivPrefix, counter);
    const cipher = createCipheriv("aes-256-gcm", this.send.key, nonce, {
      authTagLength: TAG_BYTES,
    });
    const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
    return Buffer.concat([nonce, ciphertext, cipher.getAuthTag()]);
  }

  /**
   * Verifies and decrypts a record from the device.
   *
   * Returns `null` for a record that is truncated, carries the wrong IV prefix, replays a
   * counter, or fails authentication. Callers must drop the connection rather than retry:
   * any of those means the stream is no longer trustworthy.
   */
  open(record: Buffer): Buffer | null {
    if (record.length < NONCE_BYTES + TAG_BYTES) return null;

    const nonce = record.subarray(0, NONCE_BYTES);
    const prefix = nonce.subarray(0, IV_PREFIX_BYTES);
    if (prefix.length !== this.receive.ivPrefix.length || !timingSafeEqual(prefix, this.receive.ivPrefix)) {
      return null;
    }

    const counter = readCounter(nonce);
    if (counter <= this.highestSeenReceiveCounter) return null;

    const tag = record.subarray(record.length - TAG_BYTES);
    const ciphertext = record.subarray(NONCE_BYTES, record.length - TAG_BYTES);

    try {
      const decipher = createDecipheriv("aes-256-gcm", this.receive.key, nonce, {
        authTagLength: TAG_BYTES,
      });
      decipher.setAuthTag(tag);
      const plaintext = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
      this.highestSeenReceiveCounter = counter;
      return plaintext;
    } catch {
      // Authentication failed. Deliberately indistinguishable from any other rejection.
      return null;
    }
  }
}

function buildNonce(ivPrefix: Buffer, counter: bigint): Buffer {
  const nonce = Buffer.alloc(NONCE_BYTES);
  ivPrefix.copy(nonce, 0);
  nonce.writeBigUInt64BE(counter, IV_PREFIX_BYTES);
  return nonce;
}

function readCounter(nonce: Buffer): bigint {
  return nonce.readBigUInt64BE(IV_PREFIX_BYTES);
}

/** URL-safe unpadded base64, matching `PairingSecret.encode` on the device. */
export function decodePairingSecret(encoded: string): Buffer | null {
  try {
    const buffer = Buffer.from(encoded.trim(), "base64url");
    return buffer.length === SECRET_BYTES ? buffer : null;
  } catch {
    return null;
  }
}

export function encodePairingSecret(secret: Buffer): string {
  return secret.toString("base64url");
}

export function randomSalt(): Buffer {
  return randomBytes(SALT_BYTES);
}

/**
 * Parses the `droidpilot://host:port#secret` URI shown in the Android app.
 *
 * Accepting the whole URI matters for more than convenience: it keeps the host, port and
 * secret together as one value the user copies once, instead of three fields transcribed
 * separately with the secret most likely to be truncated.
 */
export function parsePairingUri(
  uri: string,
): { host: string; port: number; secret: Buffer } | null {
  const match = /^droidpilot:\/\/([^/:]+):(\d+)#(.+)$/.exec(uri.trim());
  if (!match) return null;

  const port = Number.parseInt(match[2], 10);
  if (!Number.isInteger(port) || port < 1 || port > 65535) return null;

  const secret = decodePairingSecret(match[3]);
  if (!secret) return null;

  return { host: match[1], port, secret };
}

File: android/app/src/main/kotlin/com/mobilemcp/pro/security/SecureChannel.kt
package com.mobilemcp.pro.security

import java.security.GeneralSecurityException
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Application-layer authenticated encryption for the DroidPilot control channel.
 *
 * ## Why this and not TLS
 *
 * The channel connects two devices the same person owns, already paired out-of-band by
 * copying a secret from the phone's screen. That is a pre-shared-key problem, not a PKI
 * problem: there is no third party to vouch for, no name to verify, and no revocation
 * story. Self-signed TLS would add certificate generation, a trust-on-first-use flow and
 * a fingerprint the user has to compare — all to re-derive a shared secret the user
 * already transferred by hand.
 *
 * Equally decisive in practice: this construction is exercised by ordinary JVM unit tests
 * and by cross-implementation vectors shared with the Node client, whereas an
 * `AndroidKeyStore`-backed TLS stack can only be tested on a physical device.
 *
 * ## Construction
 *
 * Standard primitives, no novel cryptography:
 *
 *  - **Key schedule** — HKDF-SHA256 (RFC 5869). Extract over the 32-byte pairing secret
 *    with a 16-byte server-chosen per-session salt, then a single 72-byte expand split
 *    into `c2s key ‖ c2s iv ‖ s2c key ‖ s2c iv`. A fresh salt per connection means every
 *    session uses distinct keys even though the pairing secret is long-lived.
 *  - **Record protection** — AES-256-GCM. The 96-bit nonce is a 4-byte direction-specific
 *    prefix from the key schedule followed by a 8-byte big-endian record counter, the
 *    layout TLS 1.2 AEAD suites use. Counters never repeat within a session and keys never
 *    repeat across sessions, so a nonce is never reused under a key.
 *  - **Replay and reorder** — the receiver requires strictly increasing counters, so a
 *    replayed or reordered record is rejected rather than re-executed. This matters here:
 *    records are device commands, and a replayed `tap` is a real action.
 *  - **Direction separation** — client-to-server and server-to-client use different keys,
 *    so a record cannot be reflected back at its sender.
 *
 * Records are carried as binary WebSocket frames laid out as `nonce(12) ‖ ciphertext ‖ tag(16)`.
 *
 * ## What this does not defend against
 *
 * Anyone holding the pairing secret is indistinguishable from the legitimate client — the
 * secret is the entire authority. There is no forward secrecy: an attacker who records
 * traffic and later learns the secret can decrypt those recordings. Rotating the secret
 * (Settings → Regenerate) is what bounds that exposure. See SECURITY.md.
 */
class SecureChannel private constructor(
    private val sendKey: SecretKeySpec,
    private val sendIvPrefix: ByteArray,
    private val receiveKey: SecretKeySpec,
    private val receiveIvPrefix: ByteArray,
) {

    private var sendCounter: Long = 0
    private var highestSeenReceiveCounter: Long = -1

    /**
     * Encrypts [plaintext] into a self-contained record.
     *
     * Synchronised, so two callers can never be handed the same counter and therefore
     * never reuse a nonce under the same key. Note that this makes *numbering* atomic and
     * nothing more: a caller that transmits records concurrently must also serialise the
     * seal with the write, or records can reach the peer out of counter order and be
     * rejected as replays. `ControlServer.sendEncrypted` does exactly that.
     */
    @Synchronized
    fun seal(plaintext: ByteArray): ByteArray {
        val counter = sendCounter
        require(counter != Long.MAX_VALUE) { "Record counter exhausted; reconnect required" }
        sendCounter = counter + 1

        val nonce = buildNonce(sendIvPrefix, counter)
        val cipher = Cipher.getInstance(AES_GCM_TRANSFORM).apply {
            init(Cipher.ENCRYPT_MODE, sendKey, GCMParameterSpec(TAG_BITS, nonce))
        }
        val sealed = cipher.doFinal(plaintext)

        return ByteArray(NONCE_BYTES + sealed.size).also {
            nonce.copyInto(it, 0)
            sealed.copyInto(it, NONCE_BYTES)
        }
    }

    /**
     * Verifies and decrypts a record produced by the peer.
     *
     * Returns `null` for any record that fails authentication, is truncated, uses a
     * mismatched IV prefix, or replays a counter already seen. Callers must treat `null`
     * as hostile input and drop the connection rather than retrying — a failure here means
     * either corruption or an active attacker, and neither is worth a second attempt.
     */
    @Synchronized
    fun open(record: ByteArray): ByteArray? {
        if (record.size < NONCE_BYTES + TAG_BYTES) return null

        val nonce = record.copyOfRange(0, NONCE_BYTES)
        if (!MessageDigest.isEqual(nonce.copyOfRange(0, IV_PREFIX_BYTES), receiveIvPrefix)) {
            return null
        }

        val counter = readCounter(nonce)
        if (counter <= highestSeenReceiveCounter) return null

        val decrypted = try {
            Cipher.getInstance(AES_GCM_TRANSFORM).run {
                init(Cipher.DECRYPT_MODE, receiveKey, GCMParameterSpec(TAG_BITS, nonce))
                doFinal(record, NONCE_BYTES, record.size - NONCE_BYTES)
            }
        } catch (e: GeneralSecurityException) {
            // AEADBadTagException and friends: authentication failed. Deliberately opaque —
            // distinguishing failure modes here would leak information to an attacker.
            return null
        }

        highestSeenReceiveCounter = counter
        return decrypted
    }

    companion object {
        const val SECRET_BYTES = 32
        const val SALT_BYTES = 16
        private const val NONCE_BYTES = 12
        private const val IV_PREFIX_BYTES = 4
        private const val COUNTER_BYTES = 8
        private const val TAG_BITS = 128
        private const val TAG_BYTES = TAG_BITS / 8
        private const val KEY_BYTES = 32
        private const val AES_GCM_TRANSFORM = "AES/GCM/NoPadding"
        private const val HMAC_SHA256 = "HmacSHA256"

        /** Domain-separation label. Changing this is a breaking protocol change. */
        private const val HKDF_INFO = "droidpilot/v2/keys"

        private val random = SecureRandom()

        fun randomSecret(): ByteArray = ByteArray(SECRET_BYTES).also { random.nextBytes(it) }

        fun randomSalt(): ByteArray = ByteArray(SALT_BYTES).also { random.nextBytes(it) }

        /**
         * Derives a channel from the pairing [secret] and a per-session [salt].
         *
         * [isServer] selects which half of the key schedule is used for sending, so the
         * two peers derive mirrored views of the same material.
         */
        fun derive(secret: ByteArray, salt: ByteArray, isServer: Boolean): SecureChannel {
            require(secret.size == SECRET_BYTES) { "Pairing secret must be $SECRET_BYTES bytes" }
            require(salt.size == SALT_BYTES) { "Session salt must be $SALT_BYTES bytes" }

            val material = hkdf(secret, salt, HKDF_INFO.toByteArray(), 2 * (KEY_BYTES + IV_PREFIX_BYTES))
            var offset = 0
            fun take(n: Int): ByteArray = material.copyOfRange(offset, offset + n).also { offset += n }

            val c2sKey = take(KEY_BYTES)
            val c2sIv = take(IV_PREFIX_BYTES)
            val s2cKey = take(KEY_BYTES)
            val s2cIv = take(IV_PREFIX_BYTES)

            return if (isServer) {
                SecureChannel(
                    sendKey = SecretKeySpec(s2cKey, "AES"),
                    sendIvPrefix = s2cIv,
                    receiveKey = SecretKeySpec(c2sKey, "AES"),
                    receiveIvPrefix = c2sIv,
                )
            } else {
                SecureChannel(
                    sendKey = SecretKeySpec(c2sKey, "AES"),
                    sendIvPrefix = c2sIv,
                    receiveKey = SecretKeySpec(s2cKey, "AES"),
                    receiveIvPrefix = s2cIv,
                )
            }
        }

        /** HKDF-SHA256 (RFC 5869), extract-then-expand. */
        internal fun hkdf(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
            val mac = Mac.getInstance(HMAC_SHA256)
            val hashLen = mac.macLength
            require(length <= 255 * hashLen) { "HKDF output too long" }

            mac.init(SecretKeySpec(salt, HMAC_SHA256))
            val prk = mac.doFinal(ikm)

            val output = ByteArray(length)
            var previousBlock = ByteArray(0)
            var produced = 0
            var counter = 1

            while (produced < length) {
                mac.init(SecretKeySpec(prk, HMAC_SHA256))
                mac.update(previousBlock)
                mac.update(info)
                mac.update(counter.toByte())
                previousBlock = mac.doFinal()

                val chunk = minOf(hashLen, length - produced)
                previousBlock.copyInto(output, produced, 0, chunk)
                produced += chunk
                counter++
            }
            return output
        }

        private fun buildNonce(ivPrefix: ByteArray, counter: Long): ByteArray =
            ByteArray(NONCE_BYTES).also { nonce ->
                ivPrefix.copyInto(nonce, 0)
                for (i in 0 until COUNTER_BYTES) {
                    nonce[IV_PREFIX_BYTES + i] =
                        ((counter shr (8 * (COUNTER_BYTES - 1 - i))) and 0xFF).toByte()
                }
            }

        private fun readCounter(nonce: ByteArray): Long {
            var value = 0L
            for (i in 0 until COUNTER_BYTES) {
                value = (value shl 8) or (nonce[IV_PREFIX_BYTES + i].toLong() and 0xFF)
            }
            return value
        }
    }
}
